app/recurring_job.go

```diff
@@ -390,73 +311,100 @@ func (job *Job) doSnapshot() (err error) {
 		return errors.Wrapf(err, "could not get volume %v", volumeName)
 	}
 
-	if _, err := volumeAPI.ActionSnapshotCreate(volume, &longhornclient.SnapshotInput{
-		Labels: job.labels,
-		Name:   job.snapshotName,
-	}); err != nil {
+	// get all snapshot CR belong to this recurring job
+	snapshotCRList, err := job.api.Volume.ActionSnapshotCRList(volume)
+	if err != nil {
 		return err
 	}
 
-	job.logger.Infof("Created the snapshot %v", job.snapshotName)
+	snapshotCRs := filterSnapshotCRsWithLabel(snapshotCRList.Data, types.RecurringJobLabel, job.labels[types.RecurringJobLabel])
+
+	if len(snapshotCRs) > MaxRecurringJobRetain {
+		job.logger.Errorf("Trying to cleanup snapshots. There are currently too many snapshots created by this recurring job: %v snapshots. The maximum number of snapshots per recurring job is %v", len(snapshotCRs), MaxRecurringJobRetain)
+		if err := job.doSnapshotCleanup(false); err != nil {
+			return err
+		}
+		return fmt.Errorf("there are currently too many snapshots created by this recurring job: %v snapshots. The maximum number of snapshots per recurring job is %v", len(snapshotCRs), MaxRecurringJobRetain)
+	}
+
+	// Check if the latest snapshot of this recurring job is pending creation
```

Comment:
NIT: mentioning why the job will try to find the name from the existing snapshot CRs first

Reply:
NIT: abstract below codes as separate functions